### PR TITLE
Add transpose

### PR DESCRIPTION
**Pull request by gchan**
Test generator for the transpose exercise and example solution.

### CODE REVIEW

**Comment by Insti** [comment]
Super, Thanks :heart: 
I'll have a look at this later today.

File: config.json [hunk 1]
```diff
@@ -85,7 +85,8 @@
     "list-ops",
     "diamond",
     "all-your-base",
-    "isogram"
+    "isogram",
+    "transpose"
   ],
```
Comment (Insti) [inline config.json]: Added to the end, great, that is the right place to put new problems. :+1:

File: exercises/transpose/example.tt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+# Test data version:
+# <%= sha1 %>
+class TransposeTest < Minitest::Test
+<% test_cases.each do |test_case| %>
+  def <%= test_case.test_name %>
+    <%= test_case.skipped %>
+    expect = <%= test_case.expect %>
+    actual = <%= test_case.work_load %>
+    assert_equal expect, actual
```
Comment (Insti) [inline exercises/transpose/example.tt]: I really like the 3 line format you're using here. 
I'd call it `expected` but `expect` is ok too.

**Comment by Insti** [comment]
This looks good :heart:

:thinking: I wonder if there's some way of making the tests better show the transposed results.

**Comment by gchan** [comment]
Thanks @Insti, I've renamed the variable `expect` to `expected`

Could potentially use HEREDOC or string concatenation with `\` to make the transposed results look better in the test file. WDYT?

**Comment by Insti** [comment]
> Could potentially use HEREDOC or string concatenation with \ to make the transposed results look better in the test file. WDYT?

If you want to try it and see how it looks that would be good.

You could add other temporary files to the PR called `test_cases_with_heredoc.rb` or something, so we can compare the different versions. Then when we've decided which is best we can delete the redundant ones before merging.

Another thing we could do is add a comment to the test_cases.rb file asking people to submit an issue if they have an idea for a better way to display it.

**Comment by gchan** [comment]
Added files for a heredoc version for test generation. 😄

**Comment by Insti** [comment]
Ok, Thanks for doing that :heart:

The initial version is my preferred one.
I **really** don't like how the heredoc test cases end up looking.

File: exercises/transpose/transpose_test_with_heredoc.rb [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+EXPECTED
+    assert_equal expected.strip, actual
+  end
+
+  def test_simple
+    skip
+    input = <<-INPUT
+ABC
+123
+INPUT
```
Comment (kotp) [inline exercises/transpose/transpose_test_with_heredoc.rb]: With a tilde `~` in front of the Heredoc delimiter, you can present these heredocs with the proper indentation, and they will end up looking good here and clean in the generated files.  Including indentation in side the text.  This does require Ruby 2.3 though, but it is not a hard thing to write.  Facets gem has this functionality for a fairly long time (I wrote one in 2008, unsure if Facets got my version).

**Comment by kotp** [comment]
@Insti You still prefer the first version, or did the heredocs change do it for you?

**Comment by Insti** [comment]
The indented heredocs are **much** nicer.
I think we should just use the heredoc version.

File: lib/transpose_cases.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  def workload
+    'Transpose.transpose(input)'
+  end
+
+  def expect
+    [
+      "<<-EXPECTED.gsub(/^ {6}/, '')",
+      indent_lines(expected),
+      indent_line('EXPECTED')
```
Comment (Insti) [inline lib/transpose_cases.rb]: I'd extract an `indented_heredoc` method.

Comment (gchan) [inline lib/transpose_cases.rb]: Replace `indent_line('EXPECTED')` with `indented_heredoc('EXPECTED')?

Comment (Insti) [inline lib/transpose_cases.rb]: No. replace
```
"<<-EXPECTED.gsub(/^ {6}/, '')",
indent_lines(expected),
indent_line('EXPECTED')
```
with 
`indented_heredoc(expected)`

Comment (kotp) [inline lib/transpose_cases.rb]: Create a helper method to indent those things, all of them are by 6, that is the argument to the method.

`<<-EXPECTED.trim_to(6)`

Comment (gchan) [inline lib/transpose_cases.rb]: Please see latest changes. Not sure if there is a `trim_to` method for Ruby `String`?

Comment (kotp) [inline lib/transpose_cases.rb]: There could be. :)  You can make one.  That is what I was saying. The name would probably be `left_trim` though to be clear.

Comment (gchan) [inline lib/transpose_cases.rb]: For this to work I will have to monkey patch the String class with a `left_trim` method in the generated tests?

Comment (kotp) [inline lib/transpose_cases.rb]: Refinements may be an approach as well.

Comment (kotp) [inline lib/transpose_cases.rb]: Consider that this is not critical.  So this should be something to think about, and play with.  The duplication here should not stop this PR.

**Comment by Insti** [comment]
Thanks for all your work on this @gchan :heart:

**Comment by gchan** [comment]
No problem. Thanks for the feedback. I learnt a few things along the way

**Pull request merged**